optexp/datasets/wikitext103.py

```diff
@@ -0,0 +1,172 @@
+            num_workers=num_workers,
+            drop_last=False,
+            pin_memory=True,
+        )
+        return loader
+
+    def input_shape(self, batch_size: int) -> torch.Size:
+        return torch.Size([batch_size, self.sequence_length, 1])
+
+    def output_shape(self, batch_size: int) -> torch.Size:
+        return torch.Size([batch_size, self.sequence_length, self.vocab_size])
+
+    def get_num_samples(self, tr_va: TrVa) -> int:
+        tokens = self.get_tokens(tr_va, self.vocab_size)
+        n_sequences = tokens.size()[0] // self.sequence_length
+        return n_sequences * self.sequence_length
+
+    def class_counts(self, tr_va: TrVa) -> torch.Tensor:
+        tokens = self.get_tokens(tr_va, self.vocab_size)
+        n_sequences = tokens.size()[0] // self.sequence_length
+        cut_tokens = tokens[0 : n_sequences * self.sequence_length]
+        return torch.bincount(cut_tokens)
+
+    def get_dataset(self, tr_va: TrVa) -> torch.utils.data.Dataset:
+        tokens = self.get_tokens(tr_va, self.vocab_size)
+        n_tokens = tokens.size()[0]
+        n_sequences = n_tokens // self.sequence_length
+        cut_sequences = tokens[0 : n_sequences * self.sequence_length].view(
+            n_sequences, self.sequence_length
+        )
+        sequences = []
+        targets = []
+        for i in range(n_sequences):
+            sequence = cut_sequences[i]
+            target = tokens[
+                i * self.sequence_length + 1 : (i + 1) * self.sequence_length + 1
+            ]
+            sequences.append(sequence)
+            targets.append(target)
+
+        class TorchDataset(torch.utils.data.Dataset):
+            def __init__(self):
+                self.sequences = sequences
+                self.targets = targets
+
+            def __len__(self):
+                return len(self.sequences)
+
+            def __getitem__(self, idx: int):
+                return self.sequences[idx], self.targets[idx]
+
+        return TorchDataset()
+
+    def get_tokens(self, tr_va: TrVa, vocab_size: int):
+        if self.is_tokenized(tr_va):
+            tokens = torch.load(
+                open(
+                    self.directory / f"wikitext103_{tr_va}_tokenized.pt",
+                    "rb",
+                )
+            )
+            return tokens
+        raw_data = "wiki.train.tokens" if tr_va == "tr" else "wiki.valid.tokens"
+        if not self.has_tokenizer():
+            self.tokenizer.build_tokenizer(
+                self.directory / f"wikitext103_v={vocab_size}",
+                self.directory / "wiki.train.tokens",
+                vocab_size,
+            )
+        tokens = self.tokenizer.tokenize_and_numify(
+            self.directory / f"wikitext103_v={vocab_size}.model",
+            self.directory / raw_data,
+        )
+        torch.save(
+            tokens,
+            open(
+                self.directory / f"wikitext103_{tr_va}_tokenized.pt",
+                "wb",
+            ),
```

Comment:
Why is it an `open` instead of just passing the file path? If `open` is necessary it should be within a `with` statement so that it closes the file.

Reply:
I was under the impression this was necessary, as per usual I am incorrect